File: packages/arb-avm-cpp/tests/aggregator.cpp
/*
 * Copyright 2020, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "config.hpp"
#include "helper.hpp"

#include <data_storage/aggregator.hpp>
#include <data_storage/datastorage.hpp>

#include <catch2/catch.hpp>

TEST_CASE("Aggregator tests") {
    DBDeleter deleter;
    auto storage = std::make_shared<DataStorage>(dbpath);
    auto store = std::make_unique<AggregatorStore>(storage);

    SECTION("logs") {
        REQUIRE(store->logCount() == 0);
        CHECK_THROWS(store->getLog(0));
        std::vector<unsigned char> sample_log{1, 2, 3, 4};
        std::vector<char> sample_log_signed{1, 2, 3, 4};
        auto tx = storage->beginTransaction();
        store->saveLog(*tx, sample_log);
        tx = nullptr;
        /* TODO
        REQUIRE(store->logCount() == 1);
        REQUIRE(store->getLog(0) == sample_log_signed);
        CHECK_THROWS(store->getLog(1));
        */
    }

    SECTION("messages") {
        REQUIRE(store->sendCount() == 0);
        CHECK_THROWS(store->getSend(0));
        std::vector<unsigned char> sample_message{1, 2, 3, 4};
        std::vector<char> sample_message_signed{1, 2, 3, 4};
        auto tx = storage->beginTransaction();
        /* TODO
        store->saveSend(*tx, sample_message);
        tx = nullptr;
        REQUIRE(store->sendCount() == 1);
        REQUIRE(store->getMessage(0) == sample_message_signed);
        CHECK_THROWS(store->getSend(1));
        */
    }

    SECTION("requests") {
        REQUIRE(!store->getPossibleRequestInfo(10).has_value());
        store->saveRequest(10, 5);
        auto requestIndex = store->getPossibleRequestInfo(10);
        REQUIRE(requestIndex.has_value());
        REQUIRE(*requestIndex == 5);
        REQUIRE(!store->getPossibleRequestInfo(8).has_value());
    }

    SECTION("blocks") {
        CHECK_THROWS(store->latestBlock());
        std::vector<unsigned char> data{1, 2, 3, 4};
        auto tx = storage->beginTransaction();
        store->saveLog(*tx, data);
        store->saveLog(*tx, data);
        store->saveSend(*tx, data);
        store->saveSend(*tx, data);
        store->saveSend(*tx, data);
        tx = nullptr;
        std::vector<char> block_data{1, 2, 3, 4};
        store->saveBlock(50, block_data);
        {
            auto latest = store->latestBlock();
            REQUIRE(latest.first == 50);
            REQUIRE(latest.second == block_data);
        }

        tx = storage->beginTransaction();
        store->saveLog(*tx, data);
        store->saveLog(*tx, data);
        store->saveSend(*tx, data);
        tx = nullptr;
        std::vector<char> block_data2{1, 2, 3, 5};
        store->saveBlock(52, block_data2);
        {
            auto block = store->getBlock(52);
            REQUIRE(block == block_data2);
        }

        {
            // Latest is now updated
            auto latest = store->latestBlock();
            REQUIRE(latest.first == 52);
        }
    }
}

File: packages/arb-util/machine/aggregator.go
/*
 * Copyright 2019-2021, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package machine

import (
	"math/big"
	"sync"

	"github.com/ethereum/go-ethereum/core/types"
	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
	"github.com/pkg/errors"
	"github.com/rs/zerolog/log"
)

type BlockInfo struct {
	BlockLog uint64
	Header   *types.Header
}

type EVMRequestInfo struct {
	RequestId common.Hash
	LogIndex  uint64
}

type AggregatorStore interface {
	GetPossibleRequestInfo(requestId common.Hash) *uint64
	GetPossibleBlock(blockHash common.Hash) *uint64
	GetBlockInfo(height uint64) (*BlockInfo, error)
	BlockCount() (uint64, error)

	SaveBlock(header *types.Header, logIndex uint64, requests []EVMRequestInfo) error
	Reorg(height uint64) error

	CurrentLogCount() (*big.Int, error)
	UpdateCurrentLogCount(count *big.Int) error
}

type InMemoryAggregatorStore struct {
	sync.Mutex
	messages     [][]byte
	logs         []value.Value
	blocks       map[uint64]*BlockInfo
	latestBlock  uint64
	requestIndex map[common.Hash]uint64
	blockIndex   map[common.Hash]uint64
}

func NewInMemoryAggregatorStore() *InMemoryAggregatorStore {
	return &InMemoryAggregatorStore{
		blocks:       make(map[uint64]*BlockInfo),
		requestIndex: make(map[common.Hash]uint64),
		blockIndex:   make(map[common.Hash]uint64),
	}
}

func (as *InMemoryAggregatorStore) GetPossibleRequestInfo(requestId common.Hash) *uint64 {
	as.Lock()
	defer as.Unlock()
	request, ok := as.requestIndex[requestId]
	if !ok {
		return nil
	}
	return &request
}

func (as *InMemoryAggregatorStore) GetPossibleBlock(blockHash common.Hash) *uint64 {
	as.Lock()
	defer as.Unlock()
	block, ok := as.blockIndex[blockHash]
	if !ok {
		return nil
	}
	return &block
}

func (as *InMemoryAggregatorStore) GetBlock(height uint64) (*BlockInfo, error) {
	as.Lock()
	defer as.Unlock()
	blockInfo, ok := as.blocks[height]
	if !ok {
		return nil, nil
	}
	return blockInfo, nil
}

func (as *InMemoryAggregatorStore) LatestBlockInfo() (*BlockInfo, error) {
	as.Lock()
	defer as.Unlock()
	if len(as.blocks) == 0 {
		return nil, errors.New("No blocks")
	}
	return as.blocks[as.latestBlock], nil
}

func (as *InMemoryAggregatorStore) SaveBlock(header *types.Header, logIndex uint64, requests []EVMRequestInfo) error {
	as.Lock()
	defer as.Unlock()
	if logIndex >= uint64(len(as.logs)) {
		return errors.New("bad log index")
	}
	as.blocks[header.Number.Uint64()] = &BlockInfo{
		Header:   header,
		BlockLog: logIndex,
	}
	as.latestBlock = header.Number.Uint64()
	as.blockIndex[common.NewHashFromEth(header.Hash())] = header.Number.Uint64()

	for _, request := range requests {
		as.requestIndex[request.RequestId] = request.LogIndex
	}
	return nil
}

func (as *InMemoryAggregatorStore) Reorg(height uint64) error {
	as.Lock()
	defer as.Unlock()
	log.Info().
		Uint64("height", height).
		Msg("aggregator triggered reorg")
	for i := as.latestBlock; i > height; i-- {
		delete(as.blocks, i)
		if i == 0 {
			break
		}
	}
	as.latestBlock = height
	for blockHeight, block := range as.blocks {
		if blockHeight > height {
			panic("bad height")
		}
		i := block.BlockLog
		if i >= uint64(len(as.logs)) {
			panic("bad block")
		}
	}
	return nil
}

func (as *InMemoryAggregatorStore) LogCount() (uint64, error) {
	as.Lock()
	defer as.Unlock()
	return uint64(len(as.logs)), nil
}

func (as *InMemoryAggregatorStore) MessageCount() (uint64, error) {
	as.Lock()
	defer as.Unlock()
	return uint64(len(as.messages)), nil
}
